Snakemake/rules/merge_samples_vcf.smk

```diff
@@ -4,113 +4,108 @@
     run:
-        samplelist=[]
-        for file in input:
-            with open(file, 'r') as f:
-                samplelist.append(f.read().splitlines())
+#    # run in python
+        prefix = list(set([file.split('_')[0] for file in input]))
+        # group input and output files by chromosome
+
+        for name in prefix:
+        # execute for each chromosome group
+            infiles = [file for file in input if name in file] # define input files
+            outfiles = [file for file in output if name in file] # define output files
 
-        #samplelist.append(list(np.loadtxt(lst, dtype = str))) for lst in input]
-        print(len(samplelist))
+            samplelist=[] # creates holder to read in content of input files (this is updated every loop)
 
-        for a, b in itertools.combinations(samplelist, 2):
-            [b.remove(element) for element in a if element in b]
+            for ifile in infiles:  # read the content of the files and append to list (each file content is a list within list)
+                with open(ifile, 'r') as f:
+                    samplelist.append(f.read().splitlines())
 
+            for a, b in itertools.combinations(samplelist,2): # compare the lists in samplelist by pair
+                [b.remove(element) for element in a if element in b] # if id exists already in firs list, remove from second
 
-        for i, file in enumerate(output):
-            with open(file, 'w') as f:
-                for line in samplelist[i]:
-                    f.write(f'{line}\n')
-            f.close()
+            for i, ofile in enumerate(outfiles): # write out filtered lists
+                with open(ofile, 'w') as f:
+                    [f.write(f'{line}\n') for line in samplelist[i]]
+                    f.close()
 
 rule filter_samples_in_vcf:
     input:
-        files = lambda wildcards: expand(
-            ['{files}'], files = file_dic[int(wildcards.chromosome)]
-        ),
-        ids  = rules.compare_samples.output
-    output:
-        expand('Chr{{chromosome}}_{i}.vcf.gz', i = range(1, count + 1))
+        file = dir + '/{prefix}_{suffix}.vcf',
+        id = '{prefix}_{suffix}_filtered.txt'
+    output: '{prefix}_{suffix}.vcf.gz'
     shell:
         """
-        vcfs=( {input.files} )
-        ids=( {input.ids} )
-        outfiles=( {output} )
-
-        for ((i=0; i<${{#vcfs[@]}}; i++)); do
-            bcftools view -S ${{ids[$i]}} --force-samples ${{vcfs[$i]}} -O z -o ${{outfiles[$i]}}
-            bcftools index ${{outfiles[$i]}}
-        done
+        bcftools view -S {input.id} --force-samples {input.file} -O z -o {output}
+        bcftools index {output}
         """
 
 rule merge_vcfs:
     input:
-        rules.filter_samples_in_vcf.output
-        #expand('Chr{{chromosome}}_{i}.vcf.gz', i = range(1, count + 1))
-    output:
-        'Chr{chromosome}.vcf.gz'
+        files=expand('{{prefix}}_{suffix}.vcf.gz', suffix=set(SUFFIX))
```

Comment:
Isn't here only one output for multiple inputs? E.g. you merge multiple files for a chromosome into one file per chromosome?

Reply:
Is that is so, then the output should probably be just {prefix}.vcf.gz?

Reply:
Yes. But when I do like that it gives an error that it cannot determine the inputs from the outputs because they have different wildcards

Reply:
It does. It ends with one file per chromosome in the vcf folder and puts all the intermediate and initial files in separate directories in the main